grep-0001-coding-guidelines.md

```diff
@@ -0,0 +1,142 @@
+easier for multiple people to contribute to the same codebase, and will keep
+commits free of non-functional changes.
+
+To summarize, consistency is extremely useful, and coding guidelines help ensure
+consistency. We assume that all contributors are mature enough to at least try
+and follow them.
+
+## Description
+
+### General Coding Guidelines
+
+* Code is read more often than it's written. Code readability is thus something
+  worth optimizing for. All guidelines thus become secondary to better
+  readability.
+
+### General Formatting Guidelines
+
+* Never submit code with trailing whitespace.
+* Try and keep line lengths to 79 characters, unless readability suffers. We
+  often have to do fun things like SSH into machines and edit code in a
+  terminal, and do side-by-side views of code on small-ish screens, so this is
+  actually pretty helpful.
+* Always end a file with a newline.
+
+### C++-specific Guidelines
+
+* If in doubt, consult the [C++ Core Guidelines][https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md].
+  If the guidelines have an answer, and it works for you, just pick that.
+  Rationale: Many things are subjective and it's a waste of energy discussing
+  those.
+* Use Doxygen doc-blocks copiously. This makes it easier to maintain
+  auto-generated documentation.
+* All things equal, prefer standard C++ constructs over Boost constructs (see
+  also Boost guidelines). Rationale: Boost is a 3rd party dependency, and we
+  want to keep those at a minimum. Also, Boost regularly makes
+  backward-incompatible changes, making supporting a wide range of Boost
+  versions hard.
+* Given the option, prefer C++ lambdas over std::bind, and just don't use
+  boost::bind if you can. Rationale: C++ lambdas are a compiler construct, and
+  can thus be optimized better at compile time. In many cases, lambdas are
+  also more readable than the `bind` calls, which have add their own
+  syntactical idiosyncrasies (such as the `_1` argument etc.).
+* Include include files in the following order: Local headers, other GNU Radio
+  headers, 3rd-party library headers, Boost headers, standard headers.
+  The rationale is to include from most to least specific. This is the best way
+  to catch missing includes (if you were to include the standard header first,
+  it would be available to all include files that come later. If they need that
+  standard header too, they should be including it themselves).
+  Example:
+
+```cpp
+#include "module_specific.hpp"
+#include <gnuradio/block.h>
+#include <fftw3.h>
+#include <boost/shared_ptr.hpp>
+#include <mutex>
+```
+
+* Feel free to use modern C/C++ features even if they were not used before.
+  Make sure they work with the compilers and dependencies which are set for the
+  version of GNU Radio the commit will be made upon. The GNU Radio CI system
+  will be able to confirm this. (Note: C++11 features are available starting
```

Comment:
There should be a definite statement here: GNU Radio 3.7 uses C++98, from 3.8 on we use C++11.

Especially the stanza about preferring C++ constructs over boost could use that as a backing argument.

Reply:
We can have different guidelines per branch, too, with different versions. One problem that we might have is that we don't necessarily have C++11, but we have all features that are supported by the common set of C++ features between all compilers that we officially support. This is particularly true for 3.7, which actually uses a superset of C++98, but not exactly C++11.

Reply:
Just because C++11 supports something, doesn't mean MSVC gets it right (and, or, not, vs. &&, ||, !): https://github.com/gnuradio/gnuradio/pull/1571

Reply:
Exactly! Hence my reference to the CI being the golden reference...  except we don't actually test MSVC.